Chat Widget Adapters (#1098)
livechat/livechat-public-docs: src/pages/extending-chat-widget/chat-widget-adapters/index.mdx

## Available adapters

Here is a list of supported frameworks together with links to their dedicated libraries:

| Framework           | Library                                                                            |
| ------------------- | ---------------------------------------------------------------------------------- |
| **[React](#react)** | [@livechat/widget-react](https://www.npmjs.com/package/@livechat/widget-react)     |
| **[Vue](#vue)**     | [@livechat/widget-vue](https://www.npmjs.com/package/@livechat/widget-vue)         |
| **[Angular](#vue)** | [@livechat/widget-angular](https://www.npmjs.com/package/@livechat/widget-angular) |

# Universal Adapter model

Each adapter library implements the concept of integrating with Chat Widget JS API is a very similar way.
Various implemnetation detials of each framework makes them different of course but they share common lifecycle model and reactive data interface concept.

> The core implementation shared between each framework's specific adapter is distrubuted as a separate library [`@livechat/widget-core`](https://www.npmjs.com/package/@livechat/widget-core).

Reviewer: I would replace `:` with `,`. It works better for when only one item is mentioned.

> ## Component Lifecycle

Author: probably not 😅

### Mount

> During the mounting phase of the component, the Chat Widget is fully initialized. As before the mount time widget instance does not exist as also does not have access to any provided data, the mount time is the first moment when an instance of the widget is created and it receives the first set of data.

Author: no no, the Chat Widget doesn't exist and it is being initialized as a result of the `LiveChatWidget` component mount

1. Use obtained data as parameters for widget initialization

2. Create Chat Widget JS API instance similarly to a snippet code

3. Register provided event handlers for the corresponding events

> 4. Notify reactive data interface that widget has been initialized to update related data

Author: it is not a guide for the use what to do but a step by step description of the Spec which is followed by adapters implementation core

> 4. Notify reactive data interface that widget has been initialized to update related data

Reviewer: As discussed on priv, the introductory sentence could be: 

> The process of component mounting has the following steps:

### Update

> This phase allows updating provided data or callbacks by changing their value or reference. In order to do so, the component internally observes for such changes and translate them to appropriate imperative mutations via JS API method calls.

Reviewer 2: ```suggestion
This phase allows updating provided data or callbacks by changing their value or reference. In order to do so, the component internally observes such changes and translates them to appropriate imperative mutations via JS API method calls.
```

> 1. If properties were updated then call coresponding setter methods on JS API

Author: from would be better 👍

> 1. If properties were updated then call coresponding setter methods on JS API
> 
> 2. If properties that require full reload were updated then reinitialize widget with new properties
> 
> 3. If event handlers were updated then unregister the previous one and register a new one for the corresponding event

Reviewer 2: ```suggestion
1. If properties were updated then call corresponding setter methods on JS API

2. If properties that require full reload were updated then reinitialize widget with new properties

3. If event handlers were updated then unregister the previous one and register a new one for the corresponding event
```

### Unmount

> This phase results in complete Chat Widget cleanup but allowing for proper Mount phase once again.

Reviewer 2: ```suggestion
This phase results in complete Chat Widget cleanup but allows for proper Mount phase once again.
```

1. Unregister all registered callbacks

> 2. Call destroy method from JS API

Reviewer: ```suggestion
2. Call the `destroy` method from JS API.
```

Reviewer: maybe put the link here?

> 1. Unregister all registered callbacks
> 
> 2. Call destroy method from JS API
> 
> 3. Notify reactive data interface that widget has been destroyed to clear related data

Reviewer 2: ```suggestion
1. Unregister all registered callbacks

2. Call destroy method from JS API

3. Notify reactive data interface that widget has been destroyed to clear related data
```

> 3. Notify reactive data interface that widget has been destroyed to clear related data

Reviewer: ```suggestion
3. To clear the related data, notify the reactive data interface that the widget has been destroyed.
```

## Reactive data interface

> After looking into existing JS API events and available getters we can distinguish a set of entities that the Chat Widget exposes to interact with it.

Reviewer: ```suggestion
After looking into the existing JS API events and available getters, we can distinguish a set of entities that the Chat Widget exposes for interaction.
```

> Each entity contains data that may change during Chat Widget lifetime.

Reviewer 2: ```suggestion
Each entity contains data that may change during the Chat Widget lifetime.
```

> Entites are properly initialized on widget load and cleared after widget is destroyed.

Reviewer 2: ```suggestion
Entities are properly initialized on widget load and cleared after the widget is destroyed.
```

> Entites are properly initialized on widget load and cleared after widget is destroyed.

Reviewer: ```suggestion
Entites are properly initialized on widget load and cleared after widget has been destroyed.
```

Reviewer: Shows the completion of the process better.

> All posible updates are being notified for the consumer of those data.

Reviewer 2: ```suggestion
All possible updates are being notified to the consumer of that data.
```

> All posible updates are being notified for the consumer of those data.

Reviewer: The data consumer is being notified about all possible updates to that data.

> The consumer have a way to subscirbe or unsubscribe for the updates in any given moment during application lifetime.

Reviewer 2: ```suggestion
The consumer has a way to subscribe or unsubscribe for the updates at any given moment during the application lifetime.
```

> The Reactive data interface efectivly impements the Observer pattern in which a subject maintains a list of its observers and notifies them about any state changes, usually by calling one of their methods.

Reviewer 2: ```suggestion
The Reactive data interface effectively implements the Observer pattern in which a subject maintains a list of its observers and notifies them about any state changes, usually by calling one of their methods.
```

Maybe we could link a _good_ description of the observer pattern here if a reader would like to learn more quickly?

Reviewer: the reactive data interface has already been mentioned and wasn't capitalized

Author: @klarzynskik good point 👍  will look for something good 👀

Author: @klarzynskik added 😉

| **Chat** |          |
| -------- | -------- |
| chatId   | `string` |
| threadId | `string` |

| **Greeting** |          |
| ------------ | -------- |
| id           | `number` |
| uniqueId     | `string` |

| **Widget readiness** |           |
| -------------------- | --------- |
| isWidgetReady        | `boolean` |

| **Widget state** |                                        |
| ---------------- | -------------------------------------- |
| availability     | `'online' / 'offline'`                 |
| visibility       | `'maximized' / 'minimized' / 'hidden'` |

| **Customer**     |                                                  |
| ---------------- | ------------------------------------------------ |
| name             | `string`                                         |
| email            | `string`                                         |
| isReturning      | `boolean`                                        |
| isReturning      | `boolean`                                        |
| sessionVariables | `Record<string, string>`                         |
| status           | `'queued' / 'chatting' / 'browsing' / 'invited'` |

> # React adapter

Reviewer 2: Maybe we could separate the adapters' descriptions as subsections of the main "Chat Widget Adapters" section? Then, they would have their own `index.mdx` files, making it easier for us to sync them between the docs and readme directly in the package (perhaps automatically in the future). WDYT?

Author: @klarzynskik current file structure depends on Next.js file routing and that's why a single article needs to be in a single file as they are not concatenated anymore. All of our docs look like this now. The only way to separate them would be to create for each of them a dedicated article but I would like to avoid that and keep everything close together as single peace of documentation dedicated for Chat widget Adapters topic 🙂

Reviewer 2: > The only way to separate them would be to create for each of them a dedicated article

yeah, I understand the routing here, and having them separated was exactly what I was suggesting 😅 

I've meant something like that, similarly to what we have in docs e.g. for the "App Guides" (https://developers.livechat.com/docs/getting-started/app-guides):

![image](https://user-images.githubusercontent.com/47242672/152352594-565e426c-8b5d-44e3-a3c7-7db21ded72a1.png)

however, I understand your point, it's not that urgent, just thought it will be easier to sync and maintain

Author: Yes you are right I have wrongly understood your proposal 🙇‍♂️ I thought you want to "code split" the documentation as it was working on an older docs version before Next.js.

Splitting Overview and each framework adapted to different sub-article of the whole "Chat Widget Adapters" could be beneficial. I will prepare it and push on a different preview deploy hash so we will be able to compare it visually both version 👍  🙂

- [onReady](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-ready)
- [onAvailabilityChanged](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-availability-changed)
- [onVisibilityChanged](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-visibility-changed)
- [onCustomerStatusChanged](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-customer-status-changed)
- [onNewEvent](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-new-event)
- [onFormSubmitted](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-form-submitted)
- [onRatingSubmitted](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-rating-submitted)
- [onGreetingDisplayed](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-greeting-displayed)
- [onGreetingHidden](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-greeting-hidden)
- [onRichMessageButtonClicked](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-rich-message-button-clicked)

### Hooks

> This package exports a set of [React Hooks](https://reactjs.org/docs/hooks-reference.html) that allows consuming reactive data from the chat widget in any place of the application as long as the `LiveChatWidget` component is rendered in the tree.

Reviewer: The emphasize is on hooks, so plural

- [onReady](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-ready)
- [onAvailabilityChanged](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-availability-changed)
- [onVisibilityChanged](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-visibility-changed)
- [onCustomerStatusChanged](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-customer-status-changed)
- [onNewEvent](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-new-event)
- [onFormSubmitted](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-form-submitted)
- [onRatingSubmitted](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-rating-submitted)
- [onGreetingDisplayed](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-greeting-displayed)
- [onGreetingHidden](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-greeting-hidden)
- [onRichMessageButtonClicked](https://developers.livechat.com/docs/extending-chat-widget/javascript-api#on-rich-message-button-clicked)

### Services

> The `LiveChatWidgetModule`, exported from this package, registers a set of injectable services. All of them expose a subscribable [BehaviorSubject](https://rxjs.dev/api/index/class/BehaviorSubject) instance. It allows consuming reactive data from the chat widget in any place of the application, as long as the `LiveChatWidget` component is rendered in the tree.

Reviewer: ```suggestion
The `LiveChatWidgetModule`, exported from this package, registers a set of injectable services. All of them expose a subscribable [BehaviorSubject](https://rxjs.dev/api/index/class/BehaviorSubject) instance. It allows for consuming reactive data from the chat widget in any place of the application, as long as the `LiveChatWidget` component is rendered in the tree.
```
